docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,137 @@
+    - update flags for the dynamic pages
+
+## munmap 
+Calling munmap on an enclave page (dynamic allocated or not) has exactly the same effect of calling munmap on a regular RAM page. No sgx specific interface is needed. No behavior changes to current kernel space implementation.
+
+### Remarks
+
+- Enclave memory mapings are shared (MAP_SHARED). The mappings in shared processes are kept alive and independently until the process exits
+    - munmap and closing file descriptors are not required for user space. A dead process automatically releases all mappings and file descriptors. 
+- Upon all enclave mappings are removed and file handles to the enclave are closed, either by explicit munmap/fclose syscalls or when all hosting apps exited:
+    - The kernel may mark its remaining pages are reclaimable and issue EREMOVE on them any time the kernel deems appropriate.
+
+## mprotect IOCTL
+This IOCTL emulates the mprotect syscall with SGX specific extensions. In future kernel implementation, it could be mprotect or pkey_mprotect syscall with sgx extensions for the "prot" parameter.
+
+```
+#define SGX_IOC_ENCLAVE_MPROTECT _IOW(SGX_MAGIC, 0x06, struct sgx_enclave_mprotect)
+/**
+ * struct sgx_enclave_mprotect - parameter structure for the
+ *                                %SGX_IOC_ENCLAVE_MPROTECT ioctl
+ * @addr:	address of the memory to change protections
+ * @length:	length of the area.  This must be a multiple of the page size.
+ * @prot:	this must be or'ed of following:
+             PROT_READ
+             PROT_WRITE
+             PROT_EXEC
+             PROT_TRIM (new): change the page type to PT_TRIM, implies RW. User space should immediately EACCEPT, and then optionally call mprotect with PROT_EACCEPT.
+             PROT_TCS (new): change the page type to PT_TCS
+             PROT_EACCEPT (new): notify kernel EACCEPT is done for PT_TRIM type changes. This can be optional. Kernel will try EREMOVE on PT_TRIM pages in best effort on its own if user space does not notify EACCEPT.
+ */
+struct sgx_enclave_mprotect {
+	__u64	addr;
+	__u64	length;
+	__u64	prot;
+};
+```
+
+### Remarks
+
+Kernel should ensure that SGX instructions can succeed or catch and handle any fault.
+  - The kernel may maintain EPCM information on each page which includes access permission RWX, page types of PT_REG, PT_TRIM, PT_TCS.
+  - The kernel catches fault on EMODPR, EMODT and converts to error code returned to user space.
+
+The enclave run-time (or trusted run-time) may implement a parallel memory management structure which would provide information to the enclave on the enclave memory mappings.  The run-time can have a trusted API analogous to mmap which makes a call out of the enclave to issue the mmap  and then either perform EACCEPT on the pages and update the internal memory structures or configure the enclave to perform the EACCEPT when a #PF is delivered to the enclave.  With the outlined kernel interface, either implementation is possible.
```

Comment:
This is a good remark. To be honest, I think that the enclave run-time **must** implement some in-enclave memory manager. (For example, Graphene implements in-enclave VMA memory management.)

Also, I think the enclave run-time also **must** implement a rather complicated signal handling logic to correctly resolve on-demand enclave paging (to issue an `EACCEPT` instruction).

Reply:
Robust Linux sighandler in a lib/runtime is challenging :-)
A few pointers:
1) Using vDSO interface for EENTER/EXIT/AEX handling would avoid some complexity in implementing sighandler. And the vDSO interface was proposed to mitigate exactly this issue.
2) In most usages you can avoid doing EACCEPT in signal handling path. 
3) In case you need do EACCEPT in trusted side signal handler, e.g., the lazy stack expansion, implementing something similar to Intel SDK "two-phase handler" would help.